runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -325,14 +369,23 @@ public void persist() {
 
   /**
    * Extract the windows associated with the values, and invoke merge. Return a map
-   * from windows to the merge result window. If a window is not in the domain of
-   * the result map then it did not get merged into a different window.
+   * from windows to the merge result window. Windows that were not merged are present
+   * in the map with equal key and value.
    */
   private Map<W, W> collectAndMergeWindows(Iterable<WindowedValue<InputT>> values)
       throws Exception {
+    Map<W, W> windowToMergeResult = new HashMap<>();
+
     // No-op if no merging can take place
     if (windowingStrategy.getWindowFn().isNonMerging()) {
-      return ImmutableMap.of();
+      for (WindowedValue<?> value : values) {
+        for (BoundedWindow untypedWindow : value.getWindows()) {
+          @SuppressWarnings("unchecked")
+          W window = (W) untypedWindow;
+          windowToMergeResult.put(window, window);
```

Comment:
If iteration and allocation of O(<# windows>) is a concern, then you could implement a trivial `IdentityMap` here. The only windows that will be requested will be from the input, so you don't need to even check membership before returning the input to `get(W)`.